### DISCUSSION TITLE
Delete git history beyond a certain point

### DISCUSSION THREAD

**Discussion by shahzadmajeed** [newren/git-filter-repo#422] [General]
Hi, I want to delete all git history beyond a certain point such that:

1. I want to keep all commits from Jan 1, 2020 to TODAY
2. I want to keep some release branches (last 20 releases maybe), master, develop and some other currently active ones such that their parent/child relationship (for `git merge-base`) is preserved so that, on the new cleaned repo, developers can continue feature work smoothly and without confusion.
3. We don't use tags so I'm not worried about them

Is that possible with git-filter-repo? Could you please summarize on what do I have to do in steps? I really appreciate the help on this

-Shahzad

**Comment by me-and** [comment]
Here's an approach. It'll be horrendously inefficient, but it (or something like it) should work:

```
git rev-list --since 2020-01-01 master develop <list of other branches you care about> >commits-to-keep
git filter-repo --commit-callback '
  with open('commits-to-keep', 'rb') as keep_file:
    commits_to_keep = set(line.strip() for line in keep_file)
  commit.parents = [p for p in commit.parents if p in commits_to_keep]
'
```

This should edit the commits you want to keep to detach them from the commits you don't, so the commits you want to remove will be unreachable and get pruned. For less horrible inefficiency, you could use the filter-branch Python library to do something similar, but without needing to re-parse the list of commits to keep for every commit that gets processed.

Alternatively, you could find the oldest commits you want to keep, and follow the steps in the "Parent rewriting" section of the manual to rewrite them to no longer reference the unwanted parent commits. That's more manual work, but might end up being quicker.

In either case, I'm not sure "developers can continue feature work smoothly and without confusion" is likely to be achievable, per the "Discussion" section of the manual. Git deliberately makes it difficult to change or erase history, particularly for codebases that have multiple people working on them.

**Reply by shahzadmajeed** [reply]
Hi,
Thank you for the reply.

> In either case, I'm not sure "developers can continue feature work smoothly and without confusion" is likely to be achievable

I just want them to find their active branches on new repository and be able to co-relate parent/child branches like normal, thats all.

I do have the oldest commit for cutover so I will take a look at "Parent rewriting" section of the manual. Once again, I appreciate these suggestions.

**Comment by shahzadmajeed** [comment]
Hi, sorry to bother you again.

I have the oldest commit I want to keep but it is unclear from "Parent rewriting" section that how am I going to keep branches changed from "oldest commit" time to the "newest commit" time. That was the main problem I wanted to try "git-filter-repo" for. Any ideas?

Thanks again.

**Comment by me-and** [comment]
Ah, there are two "Parent rewriting" sections, which is probably not helping! The bit I'm talking about is this one: https://github.com/newren/git-filter-repo/blob/e7e245242b882d776726c1a98ce0307adbd776ff/Documentation/git-filter-repo.txt#L975-L998

Essentially, for each of the oldest commits you want to keep, run `git replace --graft <commit-hash>`. That will create a replace ref that looks like a root commit; if you run `git log` at this point, you should see that being the oldest commit in your history. The replace ref only hides the history, though; to completely remove the history you then need to call `git filter-repo --force` (the documentation explains why `--force` is needed here), which will rewrite the repository to make the replace refs permanent.

**Reply by shahzadmajeed** [reply]
Thank you for clarifying. I tried following script first to see what happens but it only kept one (top one I think) commit on each branch and discarded all other commits on all branches. I'm trying to understand why

```
git rev-list --since 2023-01-29:00:44:00 --all > commits-to-keep.txt
git filter-repo --debug --force --commit-callback '
keep_file = open("commits-to-keep.txt", "rb")
commits_to_keep = set(line.strip() for line in keep_file)
commit.parents = [p for p in commit.parents if p in commits_to_keep]
keep_file.close()
'
```

I will try parent rewriting method next...

**Reply by shahzadmajeed** [reply]
For the above script.. I wonder if `commit.parents = [p for p in commit.parents if p in commits_to_keep]` should be replaced with something like

```
commit.parents = [p for p in commit.parents if p in commits_to_keep else "NEW_ROOT_COMMIT_ID"]
```
this should be similar to grafting where "NEW_ROOT_COMMIT_ID" should connect dropped parent commits with their child commits

**Reply by me-and** [reply]
> ```
> git rev-list --since 2023-01-29:00:44:00 --all > commits-to-keep.txt
> ```

This looks like you're only trying to keep commits made in the very recent past. I would have expected you'd want something more like, say, `--since '1 January 2020'` to discard everything from 2019 and earlier, and keep everything from 2020 onwards.

**Reply by shahzadmajeed** [reply]
oh sorry I copied above script I ran for an[ experimental repository](https://github.com/shahzadmajeed/git-rewrite-history) I created yesterday. You can see the result[ this repository ](https://github.com/shahzadmajeed/git-rewrite-history-downsized). I was hoping it to keep 3 commits instead

Here is my script that I ran on a clone of my actual repo

```
# Find all commits we want to keep
cutover_date="Jan 01, 2022"
git rev-list --since="${cutover_date}" --all > commits-to-keep.txt
# For every commit we want to keep, drop all of their parents that we don't want to keep
git filter-repo --debug --force --commit-callback '
keep_file = open("commits-to-keep.txt", "rb")
commits_to_keep = set(line.strip() for line in keep_file)
commit.parents = [p for p in commit.parents if p in commits_to_keep]
keep_file.close()
'
```

but the result was same.. it only kept 1 (top) commit on each branch and deleted everything. 
EDIT: I tried switching to multiple branches and found that it actually deleted most of the code.

**Reply by shahzadmajeed** [reply]
So, I'm trying "Parent rewriting" approach now and confused about what you meant by this in your first comment:

> Alternatively, you could find the oldest commits you want to keep, and follow the steps in the "Parent rewriting" section of the manual to rewrite them to no longer reference the unwanted parent commits. That's more manual work, but might end up being quicker.

Do, I have to run `git replace --graft $oldest_commit $some_temp_root_parent` for every commit I want to keep? or just single the oldest commit that was made after the cutover point?

**Reply by me-and** [reply]
> So, I'm trying "Parent rewriting" approach now and confused about what you meant by this in your first comment:
> 
> > Alternatively, you could find the oldest commits you want to keep, and follow the steps in the "Parent rewriting" section of the manual to rewrite them to no longer reference the unwanted parent commits. That's more manual work, but might end up being quicker.
> 
> Do, I have to run `git replace --graft $oldest_commit $some_temp_root_parent` for every commit I want to keep? or just single the oldest commit that was made after the cutover point?

Say your repository looks like the below, and you want to remove all history older than the marked cut-off:

```
A     main
|
C D   feature
|/
E
|\
F G
| |   <-- cut-off
H |
|/
J     root commit
```

That means you want to replace commits `F` and `G` with ones that contain the same code but have no parents. So you'd run `git replace --graft $hashOfF` and `git replace --graft $hashOfG`. You don't specify `$some_temp_root_parent`, because you don't need any parent; you're cutting the history off beyond that point.

After running those `git replace` commands plus `git filter-repo`, the repository should look like the below:

```
A     main
|
C D   feature
|/
E
|\
F G   <- new root commits
```

**Reply by shahzadmajeed** [reply]
Yes, I tried both `replace --graft $hashOfF` and  `git replace --graft $hashOfF $some_temp_root_parent` for each commit that was going to loose a parent due to cutoff. And it looks perfect

Thank you for all the help.

**Reply by DilLip-Chowdary-Codes** [reply]
@shahzadmajeed can you share the final steps/commands for this usecase?

**Reply by shahzadmajeed** [reply]
@DilLip-Chowdary-Codes it might be different in your situation but is a high level of what I did. Note that I highly recommend trying these on a backup/clone of your repository and don't do it on original one even for final cutoff (you probably want to archive original repository as is):

```bash
# 1. Get a fresh shallowed clone since cutover date & checkout branches that fall within cutover date. Those branches will be carried over to new repo
git clone --no-single-branch --shallow-since=$cutover_date $original_repo_clone $fresh_shallowed_commits_clone_dir_name
cd $fresh_shallowed_commits_clone_dir_name
sh ../git-scripts/checkout_branches_since.sh
cd ../
 
# 2. Create another fresh but local clone from above shallowed clone so that 'git filter-repo' can work
git clone --no-single-branch "file:///$(pwd)/$fresh_shallowed_commits_clone_dir_name" $unshallowed_local_clone_dir_name
cd $unshallowed_local_clone_dir_name
 
# 3. Git Grafting - replace deleted parent commits with new parents so that their children commits are not orphans and have a root
settle_grafted_commits_for_new_root_parent
git filter-repo --force --no-ff --debug
 
# Push repository as a mirror to a new remote
git push --mirror --prune --force $new_reduced_size_repo_url
```

```bash
function settle_grafted_commits_for_new_root_parent {
    orphan_commits=($(cat .git/shallow | awk '!u[$0]++'))
    create_cutover_branch_and_new_root_commit
    new_root_commit=$(git rev-list --max-parents=0 "cutover_branch")
    git switch develop
    for orphan_child in "${orphan_commits[@]}"; do
        git replace --graft $orphan_child $new_root_commit
    done
}
```
